### PR TITLE
Draft: Optimise walkpath

### LINKED ISSUE #158: Walkpath is slow, we can make it easily 10x faster.
Related to #147 

Copying this over from some internal code @rofinn  wrote, so that this doesn't get lost and forgotten entirely.

> Slight optimization over the default `readdir(::S3Path)` based method to avoid multiple
calls when we want to grab all prefixes at once and we don't care about order:
> we may need tests or custom logic to ensure that ordering is consistent.

```julia
function _walkpath(fp::S3Path)
    # Select objects within that prefix
    objects = s3_list_objects(fp.config, fp.bucket, fp.key; delimiter="")

    # Extract keys from the results and strip the prefix out
    keys = [lstrip(replace(o["Key"], fp.key => ""), '/') for o in objects]

    # Filter any empty keys (e.g, prefix)
    filter!(!isempty, keys)

    # Return the keys joined with the original prefix
    return joinpath.(fp, keys)
end
```

Benchmark

```julia
julia> using FilePathsBase, AWSS3, BenchmarkTools

julia> fp = S3Path("REDACTED");

julia> @btime collect(walkpath($fp));
  465.464 ms (10916 allocations: 635.63 KiB)

julia> @btime collect(_walkpath($fp));
  66.557 ms (2430 allocations: 131.33 KiB)
```

### PR DESCRIPTION
**Pull request by glennmoy**
Closes #158

I think this is what we want. I've changed the original suggestion to return a `Channel` rather than a vector to align with existing behaviour. The logic for inserting the paths into the channel when `topdown = false / true` is also bit fiddly but I hope it makes sense. I can add more comments if folks like.

We can likely improve on this even more so feedback welcome.

for comparison on the speedup on a non-trivial example
```julia

# current
julia> @time collect(walkpath(path10; topdown=true));
 24.777315 seconds (164.21 k allocations: 10.317 MiB)

julia> @time collect(walkpath(path10; topdown=false));
 28.085743 seconds (164.48 k allocations: 10.400 MiB)

# this PR
julia> @time collect(walkpath(path10; topdown=true));
  0.224652 seconds (10.23 k allocations: 563.859 KiB)

julia> @time collect(walkpath(path10; topdown=false));
  0.211930 seconds (10.31 k allocations: 572.734 KiB)
```

### CODE REVIEW

File: src/s3path.jl [hunk 1]
```diff
@@ -193,6 +193,37 @@ function Base.isdir(fp::S3Path)
...
+    # Select objects with that prefix
+    objects = s3_list_objects(fp.config, fp.bucket, fp.key; delimiter="")
+
+    # Extract keys from the objects and remove the prefix
+    rm_key = s -> chop(s, head=length(fp.key), tail=0)
+    keys = map(rm_key, o["Key"] for o in objects)
+
+    isempty(keys) && return Channel{S3Path}()
+
+    chnl = S3Path[]
```
Comment (glennmoy) [inline src/s3path.jl:207]: I tried this originally via a task but 1) I couldn't check if `_p in chnl` while the `Channel` was still open and 2) I don't think I can call `reverse` on the `Channel` either so this seemed the easiest way to proceed even though it uses more memory.

**Comment by glennmoy** [comment]
2 tests from `test_walkpath` seem to be failing. Not sure what the issue is yet.

```julia
julia> result
9-element Vector{S3Path{AWS.AWSConfig}}:
 p"s3://ocaws.jl.test.20210803t192655z/pathset-root/./"
 p"s3://ocaws.jl.test.20210803t192655z/pathset-root"
 p"s3://ocaws.jl.test.20210803t192655z/pathset-root/bar/"
 p"s3://ocaws.jl.test.20210803t192655z/pathset-root/bar/qux/"
 p"s3://ocaws.jl.test.20210803t192655z/pathset-root/bar/qux/quux.tar.gz"
 p"s3://ocaws.jl.test.20210803t192655z/pathset-root/foo/"
 p"s3://ocaws.jl.test.20210803t192655z/pathset-root/foo/baz.txt"
 p"s3://ocaws.jl.test.20210803t192655z/pathset-root/fred/"
 p"s3://ocaws.jl.test.20210803t192655z/pathset-root/fred/plugh"

julia> expected
7-element Vector{S3Path{AWS.AWSConfig}}:
 p"s3://ocaws.jl.test.20210803t192655z/pathset-root/bar/"
 p"s3://ocaws.jl.test.20210803t192655z/pathset-root/bar/qux/"
 p"s3://ocaws.jl.test.20210803t192655z/pathset-root/bar/qux/quux.tar.gz"
 p"s3://ocaws.jl.test.20210803t192655z/pathset-root/foo/"
 p"s3://ocaws.jl.test.20210803t192655z/pathset-root/foo/baz.txt"
 p"s3://ocaws.jl.test.20210803t192655z/pathset-root/fred/"
 p"s3://ocaws.jl.test.20210803t192655z/pathset-root/fred/plugh"
```

**Comment by rofinn** [comment]
Yeah, I noticed that when working on an implementation that doesn't require collecting the results. FWIW, I think #179 fixes the issue and only requires a single iteration over the `s3_list_objects` results.

```julia
# The current master branch NA... gave up after 10 min :)

# gm/walkpath
julia> @time collect(walkpath(fp; topdown=false));
 46.573418 seconds (285.25 M allocations: 12.779 GiB, 3.07% gc time)

julia> @time collect(walkpath(fp; topdown=true));
 45.402200 seconds (285.14 M allocations: 12.773 GiB, 3.40% gc time)

# rf/walkpath
julia> @time collect(walkpath(fp; topdown=false));
 17.915066 seconds (1.10 M allocations: 70.752 MiB, 0.21% gc time)

julia> @time collect(walkpath(fp; topdown=true));
 19.116504 seconds (1.10 M allocations: 70.328 MiB, 0.10% gc time)
```

File: src/s3path.jl [hunk 2]
```diff
@@ -193,6 +193,37 @@ function Base.isdir(fp::S3Path)
     return iterate(objects) !== nothing
 end
 
+function FilePathsBase.walkpath(fp::S3Path; topdown=true, follow_symlinks=false, onerror=throw)
+
+    # Select objects with that prefix
```
Comment (wytbella) [inline src/s3path.jl:198]: remind me the `prefix` here refers to which parameter?

Comment (glennmoy) [reply src/s3path.jl:198]: It's the combination of the `fp.bucket` and `fp.key`, which is the path to the root directory you want to start searching in, e.g. `s3://mybucket/path/to/root`

**Pull request merged**